pkg/mkverification-raw-efi/vmcap.go

```diff
@@ -0,0 +1,348 @@
+}
+
+func (m *msr) read(index uint64, defaultVal uint64) (uint64, error) {
+	p := make([]byte, 8)
+	_, err := m.f.Seek(int64(index), 0)
+	if err != nil {
+		return defaultVal, err
+	}
+	_, err = m.f.Read(p)
+	if err != nil {
+		return defaultVal, err
+	}
+
+	data := binary.LittleEndian.Uint64(p)
+
+	return data, nil
+}
+
+type control struct {
+	name       string
+	bits       map[int]string
+	capMsr     uint64
+	trueCapMsr uint64
+}
+
+func (c *control) read2(nr uint64) (uint32, uint32) {
+	m, _ := newMsr()
+	val, _ := m.read(nr, 0)
+	return uint32(val & 0xffffffff), uint32(val >> 32)
+}
+
+func (c *control) show() {
+	fmt.Println(c.name)
+	mb1, cb1 := c.read2(c.capMsr)
+	tmb1, tcb1 := uint32(0), uint32(0)
+	if c.trueCapMsr != 0 {
+		tmb1, tcb1 = c.read2(c.trueCapMsr)
+	}
+	for bit := range c.bits {
+		zero := !(mb1&(1<<bit) != 0)
+		one := cb1&(1<<bit) != 0
+		trueZero := !(tmb1&(1<<bit) != 0)
+		trueOne := tcb1&(1<<bit) != 0
+		s := "?"
+		if c.trueCapMsr != 0 && trueZero && trueOne && one && !zero {
+			s = "default"
+		} else if zero && !one {
+			s = "no"
+		} else if one && !zero {
+			s = "forced"
+		} else if one && zero {
+			s = "yes"
+		}
+		fmt.Printf("  %-40s %s\n", c.bits[bit], s)
+	}
+}
+
+type allowedControl struct {
+	*control
+}
+
+func (a *allowedControl) read2(nr uint64) (int, uint64) {
+	m, _ := newMsr()
+	val, _ := m.read(nr, 0)
+	return 0, val
+}
+
+type misc struct {
+	name string
+	bits map[string]string
+	msr  uint64
+}
+
+func (m *misc) show() {
+	fmt.Println(m.name)
+	mymsr, _ := newMsr()
+	value, _ := mymsr.read(m.msr, 0)
+	fmt.Printf("  Hex: 0x%x\n", value)
+
+	for bits, name := range m.bits {
+		var lo, hi int
+		arr := strings.Split(bits, ",")
+		if len(arr) == 1 {
+			lo, _ = strconv.Atoi(arr[0])
+			hi, _ = strconv.Atoi(arr[0])
+		} else {
+			lo, _ = strconv.Atoi(arr[0])
+			hi, _ = strconv.Atoi(arr[1])
+		}
+
+		v := (value >> lo) & ((1 << (hi - lo + 1)) - 1)
+		var out string
+		if v == 0 {
+			out = "no"
+		} else if v == 1 {
+			out = "yes"
+		} else {
+			out = strconv.FormatUint(v, 10)
+		}
+		fmt.Printf("  %-40s %s\n", name, out)
+	}
+}
+
+func main() {
+	controls := []interface{}{
+		misc{
+			name: "Basic VMX Information",
+			bits: map[string]string{
+				"0,30":  "Revision",
+				"32,44": "VMCS size",
+				"48":    "VMCS restricted to 32 bit addresses",
+				"49":    "Dual-monitor support",
+				"50,53": "VMCS memory type",
+				"54":    "INS/OUTS instruction information",
+				"55":    "IA32_VMX_TRUE_*Ctls support",
+			},
+			msr: msrIA32VMXBasic,
+		},
+		control{
+			name: "pin-based controls",
+			bits: map[int]string{
+				0: "External interrupt exiting",
+				3: "NMI exiting",
+				5: "Virtual NMIs",
+				6: "Activate VMX-preemption timer",
+				7: "Process posted interrupts",
+			},
+			capMsr:     msrIA32VMXPinBasedCtls,
+			trueCapMsr: msrIA32VMXTruePinBasedCtls,
+		},
+		control{
+			name: "primary processor-based controls",
+			bits: map[int]string{
+				2:  "Interrupt window exiting",
```

Comment:
Ditto - is there some document we can refer to for these definitions?

Reply:
They define it in the document https://cdrdv2.intel.com/v1/dl/getContent/671200 (Appending A -VMX capability
reporting facility, on page 4483)

Reply:
Also, you can check sections 25.6.x for more information, e.g., (page 3948):

```
Table 25-5. Definitions of Pin-Based VM-Execution Controls

Bit Position(s) | Name | Description
-- | -- | --
0 | External-interrupt exiting | If this control is 1, external interrupts cause VM exits. Otherwise, they are delivered normally through the guest interrupt-descriptor table (IDT). If this control is 1, the value of RFLAGS.IF does not affect interrupt blocking.
3 | NMI exiting | If this control is 1, non-maskable interrupts (NMIs) cause VM exits. Otherwise, they are delivered normally using descriptor 2 of the IDT. This control also determines interactions between IRET and blocking by NMI (see Section 26.3).
5 | Virtual NMIs | If this control is 1, NMIs are never blocked and the “blocking by NMI” bit (bit 3) in the interruptibility-state field indicates “virtual-NMI blocking” (see Table 25-3). This control also interacts with the “NMI-window exiting” VM-execution control (see Section 25.6.2).
6 | Activate VMX- preemption timer | If this control is 1, the VMX-preemption timer counts down in VMX non-root operation; see Section 26.5.1. A VM exit occurs when the timer counts down to zero; see Section 26.2.
7 | Process posted interrupts | If this control is 1, the processor treats interrupts with the posted-interrupt notification vector (see Section 25.6.8) specially, updating the virtual-APIC page with posted-interrupt requests (see Section 30.6).
```

Reply:
Table 25-6. Definitions of Primary Processor-Based VM-Execution Controls

Bit Position(s) | Name | Description
-- | -- | --
2 | Interrupt-window exiting | If this control is 1, a VM exit occurs at the beginning of any instruction if RFLAGS.IF = 1 and there are no other blocking of interrupts (see Section 25.4.2).
3 | Use TSC offsetting | This control determines whether executions of RDTSC, executions of RDTSCP, and executions of RDMSR that read from the IA32_TIME_STAMP_COUNTER MSR return a value modified by the TSC offset field (see Section 25.6.5 and Section 26.3).
7 | HLT exiting | This control determines whether executions of HLT cause VM exits.
9 | INVLPG exiting | This determines whether executions of INVLPG cause VM exits.
10 | MWAIT exiting | This control determines whether executions of MWAIT cause VM exits.
11 | RDPMC exiting | This control determines whether executions of RDPMC cause VM exits.